src/rgw/driver/posix/bucket_cache.h

```diff
@@ -0,0 +1,549 @@
+    : bc(bc), name(name), hk(hk), flags(FLAG_NONE) {}
+
+  void set_env(std::shared_ptr<LMDBSafe::MDBEnv>& _env, LMDBSafe::MDBDbi& _dbi) {
+    env = _env;
+    dbi = _dbi;
+  }
+
+  inline bool deleted() const {
+    return flags & FLAG_DELETED;
+  }
+
+  class Factory : public cohort::lru::ObjectFactory
+  {
+  public:
+    BucketCache<D, B>* bc;
+    const std::string& name;
+    uint64_t hk;
+    uint32_t flags;
+
+    Factory() = delete;
+    Factory(BucketCache<D, B> *bc, const std::string& name)
+      : bc(bc), name(name), flags(FLAG_NONE) {
+      hk = XXH64(name.c_str(), name.length(), BucketCacheEntry::seed);
+    }
+
+    void recycle (cohort::lru::Object* o) override {
+      /* re-use an existing object */
+      o->~Object(); // call lru::Object virtual dtor
+      // placement new!
+      new (o) BucketCacheEntry(bc, name, hk);
+    }
+
+    cohort::lru::Object* alloc() override {
+      return new BucketCacheEntry(bc, name, hk);
+    }
+  }; /* Factory */
+
+  struct BucketCacheEntryLT
+  {
+    // for internal ordering
+    bool operator()(const BucketCacheEntry& lhs, const BucketCacheEntry& rhs) const
+      { return (lhs.name < rhs.name); }
+
+    // for external search by name
+    bool operator()(const std::string& k, const BucketCacheEntry& rhs) const
+      { return k < rhs.name; }
+
+    bool operator()(const BucketCacheEntry& lhs, const std::string& k) const
+      { return lhs.name < k; }
+  };
+
+  struct BucketCacheEntryEQ
+  {
+    bool operator()(const BucketCacheEntry& lhs, const BucketCacheEntry& rhs) const
+      { return (lhs.name == rhs.name); }
+
+    bool operator()(const std::string& k, const BucketCacheEntry& rhs) const
+      { return k == rhs.name; }
+
+    bool operator()(const BucketCacheEntry& lhs, const std::string& k) const
+      { return lhs.name == k; }
+  };
+
+  typedef cohort::lru::LRU<std::mutex> bucket_lru;
+
+  typedef bi::member_hook<BucketCacheEntry, member_hook_t, &BucketCacheEntry::name_hook> name_hook_t;
+  typedef bi::avltree<BucketCacheEntry, bi::compare<BucketCacheEntryLT>, name_hook_t> bucket_avl_t;
+  typedef cohort::lru::TreeX<BucketCacheEntry, bucket_avl_t, BucketCacheEntryLT, BucketCacheEntryEQ, std::string,
+			     std::mutex> bucket_avl_cache;
+
+  bool reclaim(const cohort::lru::ObjectFactory* newobj_fac) {
+    auto factory = dynamic_cast<const BucketCacheEntry<D, B>::Factory*>(newobj_fac);
+    if (factory == nullptr) {
+        return false;
+    }
+    { /* anon block */
+      /* in this case, we are being called from a context which holds
+       * A partition lock, and this may be still in use */
+      lock_guard{mtx};
+      if (! deleted()) {
+	flags |= FLAG_DELETED;
+	bc->recycle_count++;
+
+	//std::cout << fmt::format("reclaim {}!", name) << std::endl;
+	bc->un->remove_watch(name);
+#if 1
+	// depends on safe_link
+	if (! name_hook.is_linked()) {
+	  // this should not happen!
+	  abort();
+	}
+#endif
```

Comment:
well, what I was doing here was a) preferring to use normal_link; but b) have this to print info

Reply:
it's an #if iow because I'd like this normally to have normal_link; I can templatize this or something later.  I don't want asserts or ceph_asserts in the path.  if we're going to change this now, make it normal_link in the decl and just remove this block